src/Compilers/CSharp/Portable/Symbols/Metadata/PE/ExternalAnnotations.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    // PROTOTYPE(NullableReferenceTypes): external annotations should be removed or fully designed/productized
+    internal class ExtraAnnotations
+    {
+        // APIs that are useful to annotate:
+        //   1) don't accept null input
+        //   2) return a reference type
+        private static readonly ImmutableDictionary<string, ImmutableArray<ImmutableArray<bool>>> Annotations =
+            new Dictionary<string, ImmutableArray<ImmutableArray<bool>>>
+            {
+                { "System.Boolean System.Boolean.Parse(System.String)", Parameters(skip, Nullable(false)) },
+                { "System.Void System.Buffer.BlockCopy(System.Array, System.Int32, System.Array, System.Int32, System.Int32)", Parameters(skip, Nullable(false), skip, Nullable(false), skip, skip) },
+                { "System.Int32 System.Buffer.ByteLength(System.Array)", Parameters(skip, Nullable(false)) },
+                { "System.Byte System.Buffer.GetByte(System.Array, System.Int32)", Parameters(skip, Nullable(false), skip) },
+                { "System.Void System.Buffer.SetByte(System.Array, System.Int32, System.Byte)", Parameters(skip, Nullable(false), skip, skip) },
+                { "System.Byte System.Byte.Parse(System.String)", Parameters(skip, Nullable(false)) },
+                { "System.Byte System.Byte.Parse(System.String, System.IFormatProvider)", Parameters(skip, Nullable(false), skip) },
+                { "System.Byte System.Byte.Parse(System.String, System.Globalization.NumberStyles)", Parameters(skip, Nullable(false), Nullable(false)) },
+                { "System.Byte System.Byte.Parse(System.String, System.Globalization.NumberStyles, System.IFormatProvider)", Parameters(skip, Nullable(false), Nullable(false), skip) },
+                { "System.String System.String.Concat(System.String, System.String)", Parameters(Nullable(false), Nullable(true), Nullable(true)) },
+            }.ToImmutableDictionary();
+
+        internal static string MakeMethodKey(PEMethodSymbol method, ParamInfo<TypeSymbol>[] paramInfo)
```

Comment:
dumb question.  why can't hte doc-comment-ID be the method key.  Then we can reuse an existing system that already knows how to reference methods.

Reply:
This can be revised if we decide to productize the "external annotations" feature. There are many open design questions.

---
In reply to: [182234644](https://github.com/dotnet/roslyn/pull/25928#discussion_r182234644) [](ancestors = 182234644)

Reply:
>This can be revised if we decide to productize the "external annotations" feature. There are many open design questions.

Gotcha.  One of the nice things there is that at least roslyn itself can generate those strings through  our public API surface.  that way if, for example, the BCL wants to produce these 'external annotations' (perhaps for APIs that shipped in the past), then they coudl build tooling out of roslyn without having to copy this sort of code.  

--

Note: even if it isn't productized, is there a reason to generate a totally *new* symbol-description form instead of just using SymbolDisplay and picking one of hte excessively verbose options it has?  That seems like it woudl simply save you from having to define and generate these formats yourself.

Reply:
>  is there a reason to generate a totally new symbol-description form instead of just using SymbolDisplay

Yes, the main reason is that there is a circularity problem with using SymbolDisplay.
When loading a `PEMethodSymbol`, we need to load the annotations for the parameters and return type, but the symbol is not loaded. Trying to display the symbol at that time causes a cycle.